coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
+
+                anyhow::Ok(order)
+            }
+        })
+        .await??;
+
+        match order.direction {
+            commons::Direction::Short => self.shorts.add_order(order),
+            commons::Direction::Long => self.longs.add_order(order),
+        }
+        Ok(Message::NewOrder(order))
+    }
+
+    /// Matches a market order against the orderbook. Will fail if the market order can't be fully
+    /// matched.
+    async fn match_market_order(
         &mut self,
-        new_order: NewOrder,
+        new_order: NewMarketOrder,
         order_reason: OrderReason,
-    ) -> Result<Option<Message>> {
+    ) -> Result<Vec<ExecutableMatch>> {
+        let trader_pubkey = new_order.trader_id;
+        // TODO(holzeis): We might want to delay that action in a regular task allowing us to
+        // process orderbook updates much faster in memory and postpone the expensive database
+        // transactions.
         let order = spawn_blocking({
             let mut conn = self.pool.clone().get()?;
             move || {
-                let order = match new_order {
-                    NewOrder::Market(o) => {
-                        orders::insert_market_order(&mut conn, o.clone(), order_reason)
-                    }
-                    NewOrder::Limit(o) => orders::insert_limit_order(&mut conn, o, order_reason),
-                }
-                .map_err(|e| anyhow!(e))
-                .context("Failed to insert new order into DB")?;
+                let order = orders::insert_market_order(&mut conn, new_order, order_reason)
+                    .map_err(|e| anyhow!(e))
+                    .context("Failed to insert new order into DB")?;
 
                 anyhow::Ok(order)
             }
         })
         .await??;
 
-        let message = match order.order_type {
-            OrderType::Market => None,
-            OrderType::Limit => {
-                match order.direction {
-                    commons::Direction::Short => self.shorts.add(order.clone()),
-                    commons::Direction::Long => self.longs.add(order.clone()),
-                }
-                Some(Message::NewOrder(order))
+        // find a match for the market order.
+        let matched_orders = match order.direction.opposite() {
+            commons::Direction::Short => self.shorts.match_order(order.quantity),
+            commons::Direction::Long => self.longs.match_order(order.quantity),
+        };
+
+        let fail_order = {
+            let order_id = order.id;
+            let pool = self.pool.clone();
+            move || {
+                let mut conn = pool.get()?;
+                orders::set_order_state(&mut conn, order_id, OrderState::Failed)?;
+                anyhow::Ok(())
             }
         };
 
-        Ok(message)
+        let matched_quantity: Decimal = matched_orders.iter().map(|o| o.quantity).sum();
+        if matched_quantity != order.quantity {
```

Comment:
🔧 It's simpler if a failed `match_order` returns an error and does not modify the `OrderbookSide`. It's weird to have to roll back here.